core/spec/models/spree/order_shipping_method_spec.rb

```diff
@@ -0,0 +1,19 @@
+require 'spec_helper'
+
+describe Spree::Order do
+  let(:order) { create(:order_with_line_items, state: "address") }
+  let(:quadcopter) { create(:shipping_method, name: "Quads'r'us", code: "QUADSRUS", cost: 200) }
+
+  it "preserves selected shipping method when update_cart and next! called" do
+    #choose more expensive quadcopter shipping method
+    original_shipment = order.shipments.first
+    quadcopter_rate = original_shipment.add_shipping_method(quadcopter)
+    original_shipment.selected_shipping_rate_id = quadcopter_rate.id
+    expect(order.shipments.first.shipping_method).to eq quadcopter
+
+    order.contents.update_cart({}) #contents don't really matter
+    order.next!
+
+    expect(order.shipments.first.shipping_method).to eq quadcopter
+  end
+end
```

Comment:
This is the new test case that tests for the error. Updating an order followed by an `advance` which does `order.next` lost all concept of the user's prior selection of a shipping method.